fgpyo/io/__init__.py

```diff
@@ -0,0 +1,198 @@
+   >>> path_flath: Path = Path("example.txt")
+   >>> path_compressed: Path = Path("example.txt.gz")
+   >>> IO.path_is_readable(path_flat)
+   AssertionError: Cannot read non-existent path: example.txt
+   >>> IO.path_is_readable(compressed_file)
+   AssertionError: Cannot read non-existent path: example.txt.gz
+   Write to and read from path
+   >>> write_lines(path = path_flat, lines_to_write=["flat file", 10])
+   >>> write_lines(path = path_compressed, lines_to_write=["gzip file", 10])
+   Read lines from a path into a generator
+   >>> read_lines(path = path_flat)
+   ['flat file', '10']
+   >>> read_lines(path = path_compressed)
+   ['gzip file', '10']
+"""
+
+import gzip
+import io
+import os
+from pathlib import Path
+from typing import IO
+from typing import Any
+from typing import Generator
+from typing import Iterable
+from typing import Iterator
+from typing import Set
+from typing import TextIO
+from typing import Union
+from typing import cast
+
+COMPRESSED_FILE_EXTENSIONS: Set[str] = {".gz", ".bgz"}
+
+
+def assert_path_is_readable(path: Path) -> None:
+    """Checks that file exists and returns True, else raises AssertionError
+
+    Args:
+        path: a Path to check
+
+    Example:
+    >>> assert_file_exists(path = Path("some_file.csv"))
+    """
+    assert path.exists(), f"Cannot read non-existent path: {path}"
+    assert not path.is_dir(), f"Cannot read path becasue it is a directory: {path}"
+    assert os.access(path, os.R_OK), f"Path exists but is not readable: {path}"
+
+
+def assert_directory_exists(path: Path) -> None:
+    """Asserts that a path exist and is a directory
+
+    Args:
+        path: Path to check
+
+    Example:
+    >>> assert_directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def assert_path_is_writeable(path: Path, parent_must_exist: bool = True) -> Any:
+    """Asserts the following:
+    If the file exists then it must also be writeable. Else if the path is not a file and
+    parent_must_exist is true then assert that the parent directory exists and is writeable.
+    Else if the path is not a directory and parent_must_exist is false then look at each parent
+    directory until one is found that exists and is writeable.
+
+    In each scenario if a writeable file/directory is found the path is returned if not an
+    AssertionError is raised.
+
+    Args:
+         path: Path to check
+         parent_must_exist: True/False the parent directory must exist, the default is True
+
+    Example:
+    >>> assert_path_is_writeable(path = Path("example.txt"))
+    // Assuming example.txt exists and is writeable
+    >>> example.txt
+    """
+    # If file exists, it must be writeable
+    if path.is_file():
+        assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+        return path
```

Comment:
This needs to be something like:
```python
if path.exists():
    assert path.is_file() and os.access(path, os.W_OK), f"..."
```

I.e. what if the path exists, but is a directory so can't be written to?